server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -1206,7 +1231,9 @@ public ClusterState execute(ClusterState currentState) {
 
             final ImmutableOpenMap.Builder<ShardId, ShardRestoreStatus> shardsBuilder = ImmutableOpenMap.builder();
 
-            final Version minIndexCompatibilityVersion = currentState.getNodes().getMaxNodeVersion().minimumIndexCompatibilityVersion();
+            final Version minIndexCompatibilityVersion = skipVersionChecks(repositoryMetadata)
```

Comment:
We also need some special handling in allocation deciders to only restore old Lucene indices on nodes >= 8.1

Reply:
This PR is not addressing allocation-related bits, focusing on recovery bits only. There are quite a few more things to do for these indices to be properly handled at the allocation level. For example, restarting a node with a BWC index will currently fail (will detect old index and refuse to start up). This will be addressed in follow-ups.